spynnaker/pyNN/external_devices_models/spif_input_device.py

```diff
@@ -0,0 +1,296 @@
+# Copyright (c) 2021 The University of Manchester
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+from spinn_utilities.overrides import overrides
+from pacman.model.graphs.application import (
+    ApplicationFPGAVertex, FPGAConnection)
+from pacman.model.routing_info import BaseKeyAndMask
+from pacman.utilities.constants import BITS_IN_KEY
+from pacman.utilities.utility_calls import get_n_bits
+from spinn_front_end_common.abstract_models import (
+    AbstractSendMeMulticastCommandsVertex)
+from spinn_front_end_common.utilities.exceptions import ConfigurationException
+from spynnaker.pyNN.models.common import PopulationApplicationVertex
+from .spif_devices import (
+    SPIF_FPGA_ID, SPIF_OUTPUT_FPGA_LINK, SPIF_INPUT_FPGA_LINKS,
+    N_PIPES, N_FILTERS, SpiNNFPGARegister, SPIFRegister,
+    set_field_mask, set_field_shift, set_field_limit,
+    set_filter_mask, set_filter_value, set_mapper_key,
+    set_input_key, set_input_mask, set_input_route)
+
+
+class SPIFInputDevice(
+        ApplicationFPGAVertex, PopulationApplicationVertex,
+        AbstractSendMeMulticastCommandsVertex):
+    """ A 1D input device connected to SpiNNaker using a SPIF board.
+    """
+
+    #: SPIF outputs to 8 FPGA output links, so we split all input into 8
+    INPUT_MASK = 7
+
+    #: The number of devices in existence, to work out the key
+    __n_devices = 0
+
+    __slots__ = [
+        "__spif_mask",
+        "__index_by_slice",
+        "__base_key",
+        "__pipe",
+        "__key_mask",
```

Comment:
This value appears unused outside of the init.

Either use or remove from slots / self.

Reply:
Used it now!